### PR TITLE
feat(new MCP server): Add syntheticdata server

### PR DESCRIPTION
**Pull request by pranjbh**
**Issue number:**

### Changes

> Please provide a summary of what's being changed

### User experience

> Please share what the user experience looks like before and after this change

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [ ] I have reviewed the [contributing guidelines](https://github.com/awslabs/mcp/blob/main/CONTRIBUTING.md)
* [ ] I have performed a self-review of this change
* [ ] Changes have been tested
* [ ] Changes are documented

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the [project license](https://github.com/awslabs/mcp/blob/main/LICENSE).

### CODE REVIEW

File: src/syntheticdata-mcp-server/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+[project.scripts]
+"awslabs.syntheticdata-mcp-server" = "awslabs.syntheticdata_mcp_server.server:main"
+
+[dependency-groups]
+dev = [
+    "commitizen>=4.2.2",
+    "pre-commit>=4.1.0",
+    "ruff>=0.9.7",
+    "pyright>=1.1.398",
+    "pyright>=1.1.398",
```
Comment (scottschreckengaust) [inline src/syntheticdata-mcp-server/pyproject.toml]: Duplicated package, add pytest-cov 
```suggestion
    "pytest-cov>=4.1.0",
```

**Review by scottschreckengaust — CHANGES_REQUESTED** [review]

File: src/syntheticdata-mcp-server/pyproject.toml (hunk 1 above)
Comment (pranjbh) [reply src/syntheticdata-mcp-server/pyproject.toml]: Done

File: .secrets.baseline [hunk 2]
```diff
@@ -128,10 +128,12 @@
     {
       "path": "detect_secrets.filters.regex.should_exclude_file",
       "pattern": [
-        ".secrets.baseline"
+        ".secrets.baseline",
+        "src/syntheticdata-mcp-server/tests/test_constants.py",
+        "src/syntheticdata-mcp-server/tests/conftest.py"
```
Comment (scottschreckengaust) [inline .secrets.baseline]: Reviewing best methodology to exclude items.

File: src/syntheticdata-mcp-server/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
```
Comment (scottschreckengaust) [inline src/syntheticdata-mcp-server/pyproject.toml]: Consider removing duplicate
```suggestion
```

File: src/syntheticdata-mcp-server/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+[tool.pytest.ini_options]
+markers = [
+    "asyncio: marks tests that use asyncio"
+]
+asyncio_mode = "strict"
+asyncio_default_fixture_loop_scope = "function"
+
+[tool.bandit]
+exclude_dirs = [".venv", "tests"]
```
Comment (scottschreckengaust) [inline src/syntheticdata-mcp-server/pyproject.toml:120]: Consider adding coverage over the `awslabs` to exclude the `tests` directory from code coverage.
```suggestion
exclude_dirs = [".venv", "tests"]

[tool.coverage.run]
source = ["awslabs"]
```

File: .secrets.baseline [hunk 5]
```diff
@@ -128,10 +128,12 @@
...
       ]
     }
   ],
   "results": {},
```
Comment (scottschreckengaust) [inline .secrets.baseline]: Load the original `.secrets.baseline` and then `detect-secrets scan --baseline .secrets.baseline`, there will be new entries so run `detect-secrets audit .secrets.baseline` then enter you answers:

```
% detect-secrets scan --baseline .secrets.baseline # which puts a detected secret in the baseline.
% detect-secrets audit .secrets.baseline
Secret:      1 of 1
Filename:    src/syntheticdata-mcp-server/tests/test_constants.py
Secret Type: Secret Keyword
----------
8:}
9:
10:# Mock credentials using explicit test values
11:TEST_AWS_CREDENTIALS = {
12:    'AWS_ACCESS_KEY_ID': 'MOCK_KEY_123',  # Using explicit test naming
13:    'AWS_SECRET_ACCESS_KEY': 'MOCK_SECRET_123',  # Using explicit test naming
14:    'AWS_SECURITY_TOKEN': 'MOCK_TOKEN_123',  # Using explicit test naming
15:    'AWS_SESSION_TOKEN': 'MOCK_SESSION_123',  # Using explicit test naming
16:    'AWS_DEFAULT_REGION': TEST_AWS_CONFIG['region'],
17:}
----------
Should this string be committed to the repository? (y)es, (n)o, (s)kip, (q)uit: y
Saving progress...
```

The above audit adds `"is_secret": false,` to the baseline detected secret baseline.

```suggestion
        ".secrets.baseline"
      ]
    }
  ],
  "results": {
    "src/syntheticdata-mcp-server/tests/test_constants.py": [
      {
        "type": "Secret Keyword",
        "filename": "src/syntheticdata-mcp-server/tests/test_constants.py",
        "hashed_secret": "da5bcedede3c8c32004d394af1c3e14eb05b6a45",
        "is_verified": false,
        "line_number": 13,
        "is_secret": false
      }
    ]
  },
```
You can update the `generate_at` date too...

**Review by awslabs-mcp — APPROVED** [review]
# Pull Request #129 Review - Synthetic Data MCP Server

## Summary
I've reviewed the implementation of the new Synthetic Data MCP server, which provides a valuable addition to the AWS Labs MCP server ecosystem. This server enables the generation of synthetic data based on business descriptions, with facilities for data validation, saving, and loading to storage targets like S3.

## Strengths
- The implementation follows the established MCP server design guidelines and project structure
- Good test coverage at 83.93% (though there's room for improvement in a few files)
- Well-documented code with clear docstrings and type annotations
- Thorough error handling throughout the codebase
- Security considerations are well-implemented, especially in the pandas code executor
- Good modular design with clear separation of concerns

## Areas for Improvement
1. The README.md is minimal and should be expanded to match the depth of other MCP server READMEs
2. Test coverage could be improved for the missing lines identified by Codecov
3. Consider adding more examples in the documentation to show typical usage patterns

## Recommendation
I approve this PR as the code quality meets the project standards and the implementation provides a valuable new capability for the MCP server suite. Once the hold-merging label is removed and necessary approvals are obtained, this PR should be ready for merging.

---

*Note: This review was conducted following the GitHub PR review process guide for Claude Code.*

**Review by awslabs-mcp — APPROVED** [review]
# Pull Request Review: Synthetic Data MCP Server

## Summary

I've thoroughly reviewed the implementation of the new Synthetic Data MCP server that adds valuable functionality for generating, validating, and storing synthetic data for development and testing purposes. The server provides several key tools for data generation based on business descriptions, code execution in a controlled environment, and storage integration with S3.

## Code Quality Assessment

The implementation demonstrates excellent adherence to the MCP design guidelines and project structure:

1. **Architecture and Organization**
   - Well-structured modules with clear separation of concerns
   - Good modular design with proper abstraction layers
   - Comprehensive error handling throughout the codebase

2. **Security Practices**
   - Robust security checks in the pandas code executor
   - Safe evaluation of user-provided code
   - Proper validation of inputs before processing

3. **Documentation and Type Annotations**
   - Comprehensive docstrings for all classes and functions
   - Clear parameter descriptions with effective AI instructions
   - Consistent use of type annotations throughout the code

4. **Testing**
   - Good test coverage (83.93%) with comprehensive unit tests
   - Tests cover edge cases and error handling scenarios
   - Mocked external dependencies like S3 properly

## Areas for Improvement

While the implementation is generally excellent, there are a few areas that could be improved:

1. **README Documentation**
   - The README.md is minimal and should be expanded to match the depth of other MCP server READMEs
   - It would benefit from usage examples and more detailed installation instructions

2. **Test Coverage**
   - The Codecov report shows some missing coverage in a few files:
     - pandas_interpreter.py (68.81%)
     - server.py (87.05%)
     - storage/loader.py (76.47%)
     - storage/base.py (77.77%)

3. **Usability for end users**
   - Consider adding more examples in the tool documentation to show typical usage patterns
   - A few additional unit tests for edge cases would be beneficial

## Conclusion

This PR adds a well-implemented, useful MCP server that follows the established patterns and best practices. The implementation is secure, well-tested, and provides valuable functionality for generating synthetic data based on business descriptions.

I approve this PR for merging, noting that some minor documentation improvements could be made post-merge.

Run ID: 14978049930

**Review by awslabs-mcp — APPROVED** [review]
## Synthetic Data MCP Server Review

I've reviewed the new synthetic data MCP server implementation and find it to be a valuable addition to the AWS Labs MCP server ecosystem. The server provides tools for generating high-quality synthetic data based on business descriptions, validating and saving that data, and loading it to storage targets like S3.

### Strengths

- Well-structured code following the MCP server design guidelines and project patterns
- Good separation of concerns with specialized modules for different functionality
- Secure implementation of pandas code execution with proper validation
- Strong type annotations and comprehensive docstrings
- Thoughtful error handling throughout the codebase
- Support for S3 storage with partitioning and various format options
- Referential integrity checking for data validation

### Areas for Improvement

1. **Documentation**: The README.md is minimal and should be expanded with:
   - Detailed installation instructions
   - Usage examples for each tool
   - Configuration options for storage targets
   - Example business descriptions and generated data

2. **Test Coverage**: While overall coverage is good (83.93%), some files have room for improvement:
   - pandas_interpreter.py (68.81%)
   - server.py (87.05%)
   - storage/loader.py (76.47%)
   - storage/base.py (77.77%)

3. **User Experience**: 
   - Consider adding more examples in tool docstrings
   - Add documentation to help users construct effective business descriptions

### Conclusion

This PR adds a well-implemented MCP server that follows best practices and provides valuable functionality. The implementation is secure, reasonably well-tested, and adds unique capabilities to the MCP ecosystem.

I approve this PR for merging, with a suggestion to improve documentation and test coverage in future updates.

Run ID: 14981386962

**Review by awslabs-mcp — APPROVED** [review]
## PR Review Summary

I've conducted a thorough review of the new Synthetic Data MCP Server implementation, and I'm pleased to approve this pull request. This server provides valuable functionality for generating, validating, and storing synthetic data based on business use cases, which is a common requirement for development, testing, and ML model training.

### Overall Assessment

The implementation follows AWS MCP server design guidelines with well-organized code, comprehensive test coverage, and proper security measures. The code is well-structured with clear separation of concerns between data generation, validation, storage, and pandas code execution functionality.

### Key strengths of the implementation:

1. **Well-structured code organization** - Follows the MCP server design patterns with clear separation between server endpoints, data models, and storage implementations
2. **Strong security practices** - The pandas code execution functionality includes proper validation and sandboxing to prevent security issues
3. **Comprehensive test coverage** - Tests cover core functionalities, edge cases, and error scenarios
4. **Clean code style** - Code follows Python best practices and PEP 8 style guidelines
5. **Good documentation** - Tool functions include descriptive docstrings explaining functionality and usage
6. **Proper error handling** - All operations include comprehensive error handling
7. **Modular storage architecture** - Extendable storage system with S3 implementation and potential for future additions

### Minor suggestions to consider in future updates:

1. **Enhance README.md** - The current README.md is minimal and could benefit from more comprehensive documentation, similar to other MCP server implementations
2. **Consider adding more storage backends** - While S3 is well-implemented, adding additional storage backends (e.g., local file system, other cloud providers) would enhance flexibility
3. **Add more data generation templates** - Consider expanding the entity suggestions in  to cover more domain-specific entities

Overall, this is a high-quality contribution that follows the project's coding standards and provides useful functionality. The implementation is secure, well-tested, and follows MCP server design patterns.

The build checks have passed and the code structure aligns well with existing MCP servers in the repository.

Claude PR Review - Run ID: 14981894634

**Review by awslabs-mcp — DISMISSED** [review]
## PR Review Summary

I've conducted a thorough review of the new Synthetic Data MCP Server implementation, and I'm pleased to approve this pull request. This server provides valuable functionality for generating, validating, and storing synthetic data based on business use cases, which is a common requirement for development, testing, and ML model training.

### Overall Assessment

The implementation follows AWS MCP server design guidelines with well-organized code, comprehensive test coverage, and proper security measures. The code is well-structured with clear separation of concerns between data generation, validation, storage, and pandas code execution functionality.

### Key strengths of the implementation:

1. **Well-structured code organization** - Follows the MCP server design patterns with clear separation between server endpoints, data models, and storage implementations
2. **Strong security practices** - The pandas code execution functionality includes proper validation and sandboxing to prevent security issues
3. **Comprehensive test coverage** - Tests cover core functionalities, edge cases, and error scenarios
4. **Clean code style** - Code follows Python best practices and PEP 8 style guidelines
5. **Good documentation** - Tool functions include descriptive docstrings explaining functionality and usage
6. **Proper error handling** - All operations include comprehensive error handling
7. **Modular storage architecture** - Extendable storage system with S3 implementation and potential for future additions

### Minor suggestions to consider in future updates:

1. **Enhance README.md** - The current README.md is minimal and could benefit from more comprehensive documentation, similar to other MCP server implementations
2. **Consider adding more storage backends** - While S3 is well-implemented, adding additional storage backends (e.g., local file system, other cloud providers) would enhance flexibility
3. **Add more data generation templates** - Consider expanding the entity suggestions in the data generation functions to cover more domain-specific entities

Overall, this is a high-quality contribution that follows the project's coding standards and provides useful functionality. The implementation is secure, well-tested, and follows MCP server design patterns.

The build checks have passed and the code structure aligns well with existing MCP servers in the repository.

Claude PR Review - Run ID: 14981894634

**Review by awslabs-mcp — APPROVED** [review]
## Pull Request Review: Synthetic Data MCP Server

I've conducted a comprehensive review of the new Synthetic Data MCP Server implementation, which adds valuable data generation, validation, and storage functionality to the AWS MCP ecosystem.

### Summary of Findings

The implementation is well-structured, follows MCP server design guidelines, and provides robust functionality for generating synthetic data based on business descriptions, validating data structures, and integrating with storage systems.

### Key Strengths

1. **Code Quality**: The implementation demonstrates excellent adherence to project standards with properly structured modules, clear separation of concerns, and comprehensive error handling.

2. **Security Practices**: The pandas code executor includes robust security measures to prevent unsafe operations while still providing powerful data manipulation capabilities.

3. **Testing Coverage**: While there are some areas with room for improvement, the overall test coverage is good at 83.93%, with comprehensive unit tests covering key functionality.

4. **Modular Design**: The storage system is well-designed with a clear abstraction layer that allows for future extensions beyond the current S3 implementation.

5. **Documentation**: Code includes detailed docstrings and type annotations that make the codebase maintainable and understandable.

### Areas for Improvement

1. **README Documentation**: While functional, the README.md could be expanded to match the depth of other MCP server READMEs, particularly with more usage examples and detailed configuration options.

2. **Test Coverage**: Some files have room for improved test coverage, particularly:
   - pandas_interpreter.py (68.81%)
   - server.py (87.05%)
   - storage/loader.py (76.47%)
   - storage/base.py (77.77%)

3. **User Experience**: Additional examples in the documentation showing specific business scenarios would help users understand effective usage patterns.

### Conclusion

This PR introduces a valuable addition to the MCP server suite that follows project guidelines and best practices. The implementation is secure, reasonably well-tested, and provides significant functionality for data generation and management.

I approve this PR for merging, with a suggestion to enhance documentation and test coverage in future updates.

Run ID: 14983590683

**Review by jimini55 — APPROVED** [review]

**Review by scottschreckengaust — APPROVED** [review]
LGTM
